examples/vae_simple.py

```diff
@@ -0,0 +1,322 @@
+                    nonlinearity = nn.nonlinearities.sigmoid,
+                    W = nn.init.GlorotUniform() if W_dec_mu is None else W_dec_mu,
+                    b = nn.init.Constant(0.) if b_dec_mu is None else b_dec_mu,
+                    name = 'dec_output')
+            l_output_list.append(l_output)
+            if W_dec_hid is None:
+                W_dec_hid = l_dec_hid.W
+                b_dec_hid = l_dec_hid.b
+                W_dec_mu = l_output.W
+                b_dec_mu = l_output.b
+        else:
+            l_dec_mu = nn.layers.DenseLayer(l_dec_hid, num_units=x_dim,
+                    nonlinearity = None,
+                    W = nn.init.GlorotUniform() if W_dec_mu is None else W_dec_mu,
+                    b = nn.init.Constant(0) if b_dec_mu is None else b_dec_mu,
+                    name = 'dec_mu')
+            # relu_shift is for numerical stability - if training data has any
+            # dimensions where stdev=0, allowing logsigma to approach -inf
+            # will cause the loss function to become NAN. So we set the limit
+            # stdev >= exp(-1 * relu_shift)
+            relu_shift = 10
+            l_dec_logsigma = nn.layers.DenseLayer(l_dec_hid, num_units=x_dim,
+                    W = nn.init.GlorotUniform() if W_dec_ls is None else W_dec_ls,
+                    b = nn.init.Constant(0) if b_dec_ls is None else b_dec_ls,
+                    nonlinearity = lambda a: T.nnet.relu(a+relu_shift)-relu_shift,
+                    name='dec_logsigma')
+            l_output = GaussianSampleLayer(l_dec_mu, l_dec_logsigma,
+                    name='dec_output')
+            l_dec_mu_list.append(l_dec_mu)
+            l_dec_logsigma_list.append(l_dec_logsigma)
+            l_output_list.append(l_output)
+            if W_dec_hid is None:
+                W_dec_hid = l_dec_hid.W
+                b_dec_hid = l_dec_hid.b
+                W_dec_mu = l_dec_mu.W
+                b_dec_mu = l_dec_mu.b
+                W_dec_ls = l_dec_logsigma.W
+                b_dec_ls = l_dec_logsigma.b
+        l_Z_list.append(l_Z)
+    l_output = nn.layers.ElemwiseSumLayer(l_output_list, coeffs=1./L, name='output')
+    return l_enc_mu, l_enc_logsigma, l_dec_mu_list, l_dec_logsigma_list, l_output_list, l_output
+
+# ############################## Main program ################################
+
+# helper function for log-likelihood expression
+def log_likelihood(tgt, mu, ls):
+    return T.sum(-(np.float32(0.5 * np.log(2 * np.pi)) + ls)
+            - 0.5 * T.sqr(tgt - mu) / T.exp(2 * ls))
+
+def main(L=2, z_dim=2, n_hid=256, num_epochs=20, binary=True):
+    print("Loading data...")
+    X_train, X_val, X_test = load_dataset()
+    X_train_tgt = X_train.reshape(X_train.shape[0],-1)
+    X_val_tgt = X_val.reshape(X_val.shape[0],-1)
+    X_test_tgt = X_test.reshape(X_test.shape[0],-1)
+    width, height = X_train.shape[2], X_train.shape[3]
+    input_var = T.tensor4('inputs')
+    target_var = T.dmatrix('targets')
```

Comment:
1. `T.dmatrix` is a double-precision matrix, this will always end up on CPU!
2. It's a waste to have the inputs and targets twice and move them to GPU twice. You should adapt `iterate_minibatches()` to only iterate over the inputs, skip creating the `*_tgt` matrices and define the cost function only in terms of `input_var`. If you don't want the network to reshape its output back from vectors to images, you can use `input_var.flatten(2)` instead of `target_var`.